### PR TITLE
[SYCL] Implement implicit conversion for local and host accessors.

### PR DESCRIPTION
**Pull request by mmoadeli**
* Implement implicit conversion for local and host accessors.
* [Read only local accessors and implicit conversions](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:accessor.local.conversions)
* [Read only buffer host accessors and implicit conversions](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:accessor.host.buffer.conversions)
* Add tests.

### CODE REVIEW

File: sycl/include/sycl/accessor.hpp [hunk 1]
```diff
@@ -2801,6 +2804,27 @@ class __SYCL_EBO __SYCL_SPECIAL_CLASS __SYCL_TYPE(local_accessor) local_accessor
...
+
+  template <typename DataT_>
+  bool operator==(const local_accessor<DataT_, Dimensions> &Rhs) const {
+    return local_acc::impl == Rhs.impl;
+  }
+
+  template <typename DataT_>
+  bool operator!=(const local_accessor<DataT_, Dimensions> &Rhs) const {
+    return !(*this == Rhs);
+  }
```
Comment (steffenlarsen) [inline sycl/include/sycl/accessor.hpp:2826]: I would prefer if we move these to below the `using` aliases below.

File: sycl/include/sycl/accessor.hpp [hunk 2]
```diff
@@ -2709,6 +2709,9 @@ class __SYCL_SPECIAL_CLASS local_accessor_base :
   bool operator!=(const local_accessor_base &Rhs) const {
     return !(*this == Rhs);
   }
+
+private:
+  template <typename DataT_, int Dimensions_> friend class local_accessor;
```
Comment (steffenlarsen) [inline sycl/include/sycl/accessor.hpp:2714]: Why does it need to be a friend? It inherits from it and isn't `impl` a protected member?

File: sycl/include/sycl/accessor.hpp [hunk 3]
```diff
@@ -2801,6 +2804,27 @@ class __SYCL_EBO __SYCL_SPECIAL_CLASS __SYCL_TYPE(local_accessor) local_accessor
...
+                std::is_same_v<std::remove_const_t<DataT_>,
+                               std::remove_const_t<DataT>>>>
+  local_accessor(const local_accessor<DataT_, Dimensions> &other) {
+    local_acc::impl = other.impl;
+  }
+
+  template <typename DataT_>
+  bool operator==(const local_accessor<DataT_, Dimensions> &Rhs) const {
+    return local_acc::impl == Rhs.impl;
+  }
```
Comment (steffenlarsen) [inline sycl/include/sycl/accessor.hpp:2821]: Why are the inherited comparison operators not sufficient in this case?

File: sycl/include/sycl/accessor.hpp [hunk 4]
```diff
@@ -3240,6 +3264,35 @@ class __SYCL_EBO host_accessor
...
+                IsAccessReadOnly && !std::is_same_v<DataT_, DataT> &&
+                std::is_same_v<std::remove_const_t<DataT_>,
+                               std::remove_const_t<DataT>>>>
+  host_accessor(const host_accessor<DataT_, Dimensions, AccessMode> &other)
+#ifdef __SYCL_DEVICE_ONLY__
+      {}
+#else
+      : host_accessor(other.impl) {
+  }
+#endif // __SYCL_DEVICE_ONLY__
```
Comment (steffenlarsen) [inline sycl/include/sycl/accessor.hpp:3279]: Nit; I wonder how the formatter would take to
```suggestion
#ifndef __SYCL_DEVICE_ONLY__
      : host_accessor(other.impl)
#endif // __SYCL_DEVICE_ONLY__
       { }
```
Feel free to mark as resolved if the post-formatted code is absolutely ugly.

File: sycl/include/sycl/accessor.hpp (hunk 2 above)
Comment (mmoadeli) [reply sycl/include/sycl/accessor.hpp:2714]: > Why does it need to be a friend? It inherits from it and isn't `impl` a protected member?

the inheritance does not work as intended when the classes are specialised with differently.

File: sycl/include/sycl/accessor.hpp (hunk 3 above)
Comment (mmoadeli) [reply sycl/include/sycl/accessor.hpp:2821]: the inherited ones failing for the cases when `local_accessor<const>` is compared against `local_accessor<non-const>`

File: sycl/include/sycl/accessor.hpp (hunk 1 above)
Comment (mmoadeli) [reply sycl/include/sycl/accessor.hpp:2826]: I'll do it shortly.

File: sycl/include/sycl/accessor.hpp (hunk 4 above)
Comment (steffenlarsen) [reply sycl/include/sycl/accessor.hpp]: If we're going with this, it should also be applied to the one below.

File: sycl/include/sycl/accessor.hpp (hunk 2 above)
Comment (steffenlarsen) [reply sycl/include/sycl/accessor.hpp:2714]: Could we potentially just use `detail::getSyclObjImpl` instead for those cases?

Comment (mmoadeli) [reply sycl/include/sycl/accessor.hpp:2714]: great idea, I give it a try.

Comment (mmoadeli) [reply sycl/include/sycl/accessor.hpp:2714]: attempt to replace `Rhs.impl` with `detail::getSyclObjImpl(Rhs)` in both `local_accessor` and `local_accessor_base` resulted in below assert.
```
common.hpp:301:10: error: value of type 'const detail::LocalAccessorBaseDevice<AdjustedDim>' is not contextually convertible to 'bool'
  assert(SyclObject.impl && "every constructor should create an impl"); 
```

**Review by steffenlarsen — APPROVED** [review]
LGTM!

**Pull request merged**